events-wfm/index.js

```diff
@@ -0,0 +1,236 @@
+      const endAtSecond = moment.unix(availableEndAtForChannel).second();
+      const endAtInterval = (endAtMinuteRemainder === TIME_BUCKET_IN_MINUTES && endAtSecond === 0) ? availableEndAtForChannel : availableEndAtForChannel + endAtMinuteRemainder*60 - endAtSecond;
+
+      const numIntervalsInBetween = (endAtInterval-startAtInterval) / (TIME_BUCKET_IN_MINUTES * 60);
+
+      for(var i = 0; i < numIntervalsInBetween; i++) {
+        let thisIntervalStartAt = startAtInterval + i*TIME_BUCKET_IN_MINUTES*60;
+        let thisIntervalEndAt = thisIntervalStartAt + TIME_BUCKET_IN_MINUTES*60;
+        let niceFormatIntervalStartAtIdentifier = moment.unix(thisIntervalStartAt).tz(TIMEZONE).format('YYYY-MM-DD HH:mm');
+
+        let agentRawReportRowIdentifier = `${thisAgentId}-${niceFormatIntervalStartAtIdentifier}-${channel}`;
+
+        if(typeof agentRawReport[agentRawReportRowIdentifier] === 'undefined') {
+          agentRawReport[agentRawReportRowIdentifier] = {
+            'Agent ID': thisAgentId,
+            'Interval Start At': niceFormatIntervalStartAtIdentifier,
+            'Timezone': TIMEZONE,
+            'Interval Duration (minute)': TIME_BUCKET_IN_MINUTES,
+            'Available Time (minutes)': 0,
+            'Channel': channel
+          }
+        }
+
+        //Agent was available for the entire interval
+        if(thisIntervalStartAt >= availableStartAtForChannel && thisIntervalEndAt <= availableEndAtForChannel) {
+          agentRawReport[agentRawReportRowIdentifier]['Available Time (minutes)'] += TIME_BUCKET_IN_MINUTES;
+        }
+        //Agent was available at the beginning of the interval, up until when they changed availability
+        else if (thisIntervalStartAt >= availableStartAtForChannel && thisIntervalEndAt > availableEndAtForChannel) {
+          agentRawReport[agentRawReportRowIdentifier]['Available Time (minutes)'] += (availableEndAtForChannel-thisIntervalStartAt)/60.0;
+        }
+        //Agent was available after the beginning of this interval, up until the end of the interval
+        else if (thisIntervalStartAt < availableStartAtForChannel && thisIntervalEndAt <= availableEndAtForChannel) {
+          agentRawReport[agentRawReportRowIdentifier]['Available Time (minutes)'] += (thisIntervalEndAt -availableStartAtForChannel)/60.0;
+        }
+        //Agent was available in between this interval
+        else if (thisIntervalStartAt < availableStartAtForChannel && thisIntervalEndAt > availableEndAtForChannel) {
+          agentRawReport[agentRawReportRowIdentifier]['Available Time (minutes)'] += (availableEndAtForChannel -availableStartAtForChannel)/60.0;
+        }
+      }
+    }
+  }
+
+  let ret = [];
+  for(i in agentRawReport) {
+    ret.push(agentRawReport[i]);
+  }
+
+  return ret;
+}
+
+//This function works with the exception of when the first Hold event for the agent
```

Comment:
This wording here is a bit confusing. I tried to clarify the way I understood it. Feel free to ignore :) 

This function works around an issue with the Agent Events API and first hold events which start on the previous day. As an example, if we pull Agent events using the API for the time period of 00:00 AM - 00:30 AM and an agent went on hold at 11:59 PM the previous day, that agent would not be included in the event results.